CONTRIBUTING.md

```diff
@@ -232,3 +232,118 @@ error_session:
 	return NULL;
 }
 ```
+
+## Wayland protocol implementation
+
+Each protocol generally lives in a file with the same name, containing one
+struct per interface. For instance, `xdg_shell` lives in `types/xdg_shell.h` and
+has a `wlr_xdg_surface` struct.
+
+### Globals
+
+Global interfaces generally have public constructors and destructors. Their
+struct has a field holding the `wl_global` itself, a list of resources clients
+created by binding to the global, a destroy signal and a `wl_display` destroy
+listener. Example:
+
+```c
+struct wlr_compositor {
+	struct wl_global *wl_global;
+	struct wl_list wl_resources;
+	…
+
+	struct wl_listener display_destroy;
+
+	struct {
+		struct wl_signal new_surface;
+		struct wl_signal destroy;
+	} events;
+};
+```
+
+When the destructor is called, it should trigger the destroy signal, remove the
+display destroy listener, destroy the `wl_global`, destroy all bound resources
+and then destroy the struct.
+
+### Resources
+
+Resources have a user data field which can be access through
+`wl_resource_get_data`. To prevent bad casts, a safe helper function checking
+the type of the resource is used:
+
+```c
+static const struct wl_surface_interface surface_impl;
+
+struct wlr_surface *wlr_surface_from_resource(struct wl_resource *resource) {
+	assert(wl_resource_instance_of(resource, &wl_surface_interface,
+		&surface_impl));
+	return wl_resource_get_user_data(resource);
+}
+```
+
+### Destructing resources
+
+Resource structs should only be destroyed when their resource is destroyed, ie.
+in the resource destroy handler (set with `wl_resource_set_implementation`).
+Destructor requests should only call `wl_resource_destroy`.
+
+### Inert resources
+
+Some resources can become inert according to the protocol. All requests made to
+inert resources should be ignored, except the destructor. This is achieved by:
+
+1. When the resource becomes inert: destroy the struct and call
+   `wl_resource_set_user_data(resource, NULL)`. Do not destroy the resource.
+2. For each request made to a resource that can be inert: add a NULL check to
+   ignore the request if the resource is inert.
+3. When the client calls the destructor request on the resource: call
+   `wl_resource_destroy(resource)` as usual.
+4. When the resource is destroyed, if the resource isn't inert, destroy the
+   struct.
```

Comment:
Another reason a resource could be inert is if the compositor wants to get rid of it. Since only clients can destroy resources we're kind of SoL